[EDNA-43] Implement Upload service fully

gromakovsky opened the pull request:
## Description

1. Implemented `compoundNameToId` and `targetNameToId` used in `parseFile`, so now it's fully implemented.
2. Implemented `uploadFile`.
3. Added `UploadFile` to state machine tests.

## Related issue(s)

Resolves https://issues.serokell.io/issue/EDNA-43

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

gromakovsky committed:
Add tests for uploadFile

Problem: we are going to implement uploadFile and we need tests for it.
Solution: add UploadFile action to state machine tests.

gromakovsky committed:
Implement parseFile fully

gromakovsky committed:
Implement uploadFile

Problem: `uploadFile` has a stub implementation and doesn't do what it
should.
Solution: implement it properly.
Some changes have been made to DB schema: creation dates now have
default values (`now()`), metadata is not nullable anymore
because we can use empty list or JSON's `null` when it's empty.
Same for `analysis_method.parameters`.
Also we create one analysis_method by default because we need just
one method.

gromakovsky committed:
Introduce EDNA_DEBUG_DB

Problem: sometimes it's convenient to see all SQL requests made
by edna-server. Currently it's only possible if you modify some
code.
Solution: to facilitate this process `edna-server` now reads
`EDNA_DEBUG_DB` env var and enables debug logging (using
`runBeamPostgresDebug`) if it's set to 1 or True.

gromakovsky committed:
Add unit tests for the Upload service

Problem: we have state machine tests for upload service, but they are
quite complex by design and it's not always obvious whether the right
checks are executed.

Solution: add simple unit tests for this service that operate with
concrete data.

gromakovsky commented:
There is a problem with the current state machine tests. When a new target is added in `applyUploadFile`, we assign its ID as `max_id + 1`. But actual implementation may assign a different ID. When it does `INSERT INTO targets`, `target_id` will be incremented by 1 even in case of unsuccessful insert (that may if we insert a target that is already added). So we can easily have 2 targets with IDs 1 and 10 in the DB, but in our model we assume that all IDs are consecutive.
So instead of assigning IDs ourselves we need to use IDs provided by actual implementation. We can take them from the result of `UploadFile` command, but inside `applyUploadFile` this result is available as `Var Result v`. And apparently it's impossible to get anything from this `Var Result`, we can't get `Var PartOfResult v` for example. So the only thing we can do is to put this `Var` into `EdnaState`. I tried to do that, but it gave a lot of complexity and I decided to postpone it.

I've also added some unit tests. We can't spend too much time on SMT at this stage, let's try to improve them later.

gromakovsky committed:
Fix handling of ServantError

Problem: in `Edna` monad we sometimes throw `ServantError`.
Servant engine doesn't handle `ServantError` if its thrown using
builtin Haskell's exception machinery, it uses `ExceptT` instead.
So when we throw `err501` for example, it's treated as an arbitrary
Haskell exception and results into 500 HTTP code (internal error).

Solution: update `ednaToHandler` to consider `ServantError` in
addition to `EdnaServerError`.

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -30,12 +43,22 @@ parseFile content =
 compoundNameToId :: Text -> Edna (Maybe (SqlId Compound))
-compoundNameToId _ = pure Nothing
+compoundNameToId compoundName =
+  fmap (fmap mkSqlId) . runSelectReturningOne' $ select $ do
```
Maybe use `<<$>>`?

sashasashasasha151 replied:
Why don't you like `<<$>>`?

gromakovsky replied:
Oh, right. Well, I don't like it for a couple of reasons: it's non-standard (people who are not used to `Universum` don't know this operator) and I always forget whether it's `<<$>>` or `<$$>`. And here there is an additional reason: it can't be used as an operator in infix position, it can be used only as a function and I slightly dislike that as well. So I think `fmap (fmap …` is ok here.

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -3,19 +3,32 @@
 import Edna.ExperimentReader.Parser (parseExperimentXls)
-import Edna.ExperimentReader.Types
+import Edna.ExperimentReader.Types as EReader
 import Edna.Setup
-import Edna.Web.Types
+import Edna.Web.Types hiding (cName, tName)
```
Are implicit imports match Serokell codestyle?

gromakovsky replied:
Yes, implicit imports from the current project are permitted.

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+data UploadError =
+    UEUnknownProject (SqlId Project)
+  | UEUnknownTestMethodology (SqlId TestMethodology)
+  deriving stock (Show, Eq)
```
I think Errors should be in Errors module

sashasashasasha151 commented on backend/README.md:
```diff
@@ -19,3 +19,6 @@ Then run `make test` to execute tests.
 * The server endpoints will be available at `http://localhost:9000/api/*`.
 * Swagger docs will be available at `http://localhost:9000/docs/`.
+
+Set `EDNA_DEBUG_DB=1` to enable logging of all DB actions made.
```
Maybe put it to config?

gromakovsky replied:
I have several reasons why it's done this way, but the main one is that I need it to also work in tests and it's hard (maybe impossible) to modify config in tests without recompiling code.

gromakovsky replied:
If you mean "put it to config AND keep it as env variable", it probably makes sense, but it's extra work which is not needed now, we can do it in [EDNA-44](https://issues.serokell.io/issue/EDNA-44) when we have time.

sashasashasasha151 commented on backend/sql/init.sql:
```diff
@@ -132,9 +132,11 @@ create table if not exists analysis_method
+    parameters         json   not null
 );
 
+insert into analysis_method values (default, 'IC50', '[]'::json);
```
In dev mode (when we run this schema on each start without reset) it will insert in every time. I think you can write simple function here, which will check if `analysis_method` is empty or not and insert value

gromakovsky replied:
Oh right, but I guess I don't need a function here, I can do `insert into analysis_method values (1, 'IC50', '[]'::json) on conflict do nothing`.

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+
+      measurementsToSummary fileMeasurements
+
+insertTarget :: Text -> Edna (Text, Word32)
+insertTarget targetName = (targetName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esTarget ednaSchema)
+    (insertExpressions [TargetRec default_ (val_ targetName) default_])
+    (Pg.onConflict (Pg.conflictingFields tName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added target not found: " <> targetName)  <$>
+    targetNameToId targetName
+
+insertCompound :: Text -> Edna (Text, Word32)
+insertCompound compoundName = (compoundName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esCompound ednaSchema)
+    (insertExpressions [CompoundRec default_ (val_ compoundName) default_ default_])
+    (Pg.onConflict (Pg.conflictingFields cName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added compound not found: " <> compoundName)  <$>
+    compoundNameToId compoundName
```
You can use `runInsertReturningOne` just here without calling `targetNameToId`.

sashasashasasha151 replied:
It is also possible to just run `runInsertReturningList` with `onConflictDoNothing` and get all new names

gromakovsky replied:
No, I can't use `runInsertReturningOne` because it returns nothing if this compound already exists. We [discussed](https://serokell.slack.com/archives/C01F8AQSHGE/p1616340241019200) it yesterday.

sashasashasasha151 replied:
But what about `runInsertReturningList`?

gromakovsky replied:
It will return empty list if nothing is updated and I will need to do `SELECT` after that. I can do the following: `INSERT RETURNING`, check whether anything was returned, and if not, call `SELECT`. This way when we add a new target/compound we will execute one command instead of two. But there will be more code in this case and I don't think we'll ever notice this optimization, it's very minor.

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+insertExperimentFile ::
+  Word32 -> Word32 -> FileMetadata -> Text -> Text -> LByteString -> Edna Word32
+insertExperimentFile projId methodId meta descr fileName blob = do
+  unSerial . expectOneInsertion . map efExperimentFileId <$> runInsertReturningList'
```
`runInsertReturningOne` looks more convenient

gromakovsky replied:
Interesting, I don't know why I did it this way, probably because there is no `runInsertReturningOne'` in `Edna.DB.Integration`, but I guess I can add it.

gromakovsky replied:
Apparently there is no `runInsertReturningOne` in `beam`. But I still can define it.

sashasashasasha151 replied:
There is `runInsertReturning`, isn't it what you want?

sashasashasasha151 replied:
Looks like it isn't (

gromakovsky replied:
It looks a little bit different. Anyway, I've defined
```
-- | Insert items and expect one item in the result. The caller is responsible
-- for ensuring that exactly one result will be returned.
runInsertReturningOne' ::
  ( Beamable table
  , FromBackendRow Postgres (table Identity)
  ) => SqlInsert Postgres table -> Edna (table Identity)
runInsertReturningOne' = fmap expectOneInsertion . runInsertReturningList'
  where
    expectOneInsertion :: HasCallStack => [x] -> x
    expectOneInsertion = \case
      [x] -> x
      xs -> error $
        "Expected to insert 1 item, but inserted: " <> show (length xs)
```

and it's probably sufficient for now

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+    (insert (esExperiment ednaSchema) $ insertExpressions [experimentRec])
+  subExpId <-
+    unSerial . expectOneInsertion . map seSubExperimentId <$> runInsertReturningList'
+    (insert (esSubExperiment ednaSchema) $ insertExpressions [subExperimentRec expId])
```
Maybe `runInsertReturningOne`

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+  subExpId <-
+    unSerial . expectOneInsertion . map seSubExperimentId <$> runInsertReturningList'
+    (insert (esSubExperiment ednaSchema) $ insertExpressions [subExperimentRec expId])
+  forM_ measurements $ \measurement -> do
+    let
+      measurementRec :: MeasurementT (QExpr Postgres s)
+      measurementRec = MeasurementRec
+        { mMeasurementId = default_
+        , mExperimentId = val_ expId
+        , mConcentration = val_ (EReader.mConcentration measurement)
+        , mSignal = val_ (EReader.mSignal measurement)
+        , mIsOutlier = val_ (EReader.mIsOutlier measurement)
+        }
+    measurementId <-
+      unSerial . expectOneInsertion . map mMeasurementId <$> runInsertReturningList'
+      (insert (esMeasurement ednaSchema) $ insertExpressions [measurementRec])
+    when (EReader.mIsOutlier measurement) $
+      runInsert' $ insert (esRemovedMeasurements ednaSchema) $ insertValues
+      [RemovedMeasurementsRec subExpId measurementId]
```
I am not sure that I am right, but as I understand call sql `1` time is better than calling it `n` times. So, you can construct list of `measurementRec` and insert them using `runInsertReturningList` and get it's `ids` with `isOutliers`, filter on `isOutliers` and again construct a list of sub-experiments and simply call sql one time

gromakovsky replied:
Ok, sounds doable. I suppose there will be 2 SQL calls instead of this `forM_`, not 1.

sashasashasasha151 commented on backend/src/Edna/Upload/Service.hs:
```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+expectOneInsertion :: (HasCallStack, Show x) => [x] -> x
+expectOneInsertion = \case
+  [x] -> x
+  xs -> error $ "Expected to insert 1 item, but inserted: " <> show xs
```
Why not `specific` errors?

gromakovsky replied:
What is `specific`? Anyway, this function will likely be removed because you say there is `runInsertReturningOne`.

sashasashasasha151 replied:
So, create smth like `data DBError = `. There are also other plain errors in this module

gromakovsky replied:
Because this is a programmer mistake which should never happen. I generally follow [these guidelines](https://github.com/input-output-hk/cardano-sl/blob/2a5b5769799fd220e2ff6a6b09a1832d8e6f06c0/docs/exceptions.md). `data DBError` is useful when there is a chance that this error will be handled somehow because you can turn it into some other format (e. g. HTTP response) or even translate to another language. But this error is a programmer mistake which we don't need to handle specially.

gromakovsky committed:
Reorganize Upload errors

Problem: upload error is defined inside Service.hs and increases its
size. Also some possible errors are defined in Edna.Web.Types. There
we have EdnaServerError which is supposed to group all errors, but
we are going to use structure of modules where each domain has
Edna.Domain modules and thus it makes more sense to have errors per
domain. There is no need to group them into one EdnaServerError type
because such grouping is already done by means of built-in exception
machinery.

Solution:
1. Create Edna.Upload.Error with all upload-specific errors.
2. Move EdnaServerError there and rename to UploadApiError.
3. Delete `XlsxParingError` because it's used only in a legacy
function and is not needed there.
4. Provide a default definition for `toServerError` to make it easier
to define its instances.

gromakovsky committed:
Consider all application errors in `ednaToHandler`

Problem: `ednaToHandler` is supposed to map application errors to HTTP
responses represented as `ServerError`. Currently it does so not for
all errors.

Solution: consider 2 more types of errors there.

gromakovsky commented:
@sashasashasasha151 I've accidentally squashed 2 fixup commits, but you still can check them here:
* https://github.com/serokell/edna/commit/03f53786273b78a7b613e57550244db42b21bdfd
* https://github.com/serokell/edna/commit/1e121b0430512e24c1332e8bacbbcb2b3f772cf4

The last thing I need to do here is to update `FileSummary` to always contain names. I'll do it shortly.

gromakovsky committed:
Always return all names from Upload endpoints

Problem: currently Upload endpoints return names only if IDs are not
available. It may cause frontend to make extra calls just to get names
for IDs.
Solution: to make frontend's life easier we return names for all items
in `FileSummary`. To minimize the number of changes we simply change
`Either Id Name` to `(Maybe Id, Name)`.

gromakovsky commented:
@sashasashasasha151 you are welcome to re-review.

sashasashasasha151 commented:
> @sashasashasasha151 you are welcome to re-review.

I have email notifications. So, you can just rerequest my review

gromakovsky committed:
Change /file/parse to POST

Problem: `/file/parse` endpoint currently can't be executed because
it's defined as GET, but multipart apparently must be POST, or at
least can't be GET and HEAD.

Solution: change it to POST.

gromakovsky committed:
Drop ReqBody from `/file/upload`

Problem: servant's `MultipartForm` doesn't work with `ReqBody` in one
endpoint and that's what we are trying to do in `/file/upload`.

Solution: stop using `ReqBody`, make all 3 fields Captures.
